### PR TITLE
refactor(core-processor, core): Reimburse reduced for the reservation gas, if it was unreserved immediately

### PR DESCRIPTION
**Pull request by techraed**
Basically returns to the `GasAmount.left` the gas that was reduced to create a new `GasReservationState::Created` reservation.

### CODE REVIEW

File: core-processor/src/ext.rs [hunk 1]
```diff
@@ -1199,15 +1199,37 @@ impl<LP: LazyPagesInterface> Externalities for Ext<LP> {
...
+                .costs
+                .rent
+                .reservation
+                .cost_for(
+                    self.context
+                        .reserve_for
+                        .saturating_add(reimbursement.duration())
+                        .into(),
+                )
+                .saturating_add(amount);
```
Comment (ark0f) [inline core-processor/src/ext.rs:1219]: This and the same code in `reserve_gas` can be moved into 1 function

File: core-processor/src/ext.rs [hunk 2]
```diff
@@ -1199,15 +1199,37 @@ impl<LP: LazyPagesInterface> Externalities for Ext<LP> {
...
+            self.context
+                .gas_counter
+                .increase(reimbursement_amount, reimbursement)
+                .then_some(())
+                .unwrap_or_else(|| {
+                    unreachable!(
```
Comment (ark0f) [inline core-processor/src/ext.rs:1225]: I guess `log::error!()` should be here too?

File: core-processor/src/ext.rs [hunk 3]
```diff
@@ -2182,4 +2211,94 @@ mod tests {
...
+        // Unreserving existing and checking no gas reimbursed.
+        let gas_before = ext.gas_amount();
+        assert!(ext.unreserve_gas(id).is_ok());
+        let gas_after = ext.gas_amount();
+
+        assert_eq!(gas_after.left(), gas_before.left());
+    }
+
+    #[test]
+    fn test_unreserve_with_reimbursement() {
```
Comment (ark0f) [inline core-processor/src/ext.rs:2268]: `CREATE + USE + TRY UNRESERVE` case is required

**Review by ark0f — APPROVED** [review]

File: core-processor/src/ext.rs [hunk 4]
```diff
@@ -2182,4 +2211,94 @@ mod tests {
...
+        let id = ext
+            .reserve_gas(reservation_amount, duration)
+            .expect("internal error: failed reservation");
+        let gas_before = ext.gas_amount();
+        assert!(ext.unreserve_gas(id).is_ok());
+        let gas_after = ext.gas_amount();
+
+        assert_eq!(
+            gas_after.left(),
+            gas_before.left() + reservation_amount + duration_cost
```
Comment (breathx) [inline core-processor/src/ext.rs:2301]: and eq gas at the very beginning

**Review by breathx — APPROVED** [review]

File: core-processor/src/ext.rs (hunk 3 above)
Comment (techraed) [reply core-processor/src/ext.rs:2268]: yup, covered

File: core-processor/src/ext.rs (hunk 1 above)
Comment (techraed) [reply core-processor/src/ext.rs]: done

File: core-processor/src/ext.rs (hunk 2 above)
Comment (techraed) [reply core-processor/src/ext.rs]: done

File: core-processor/src/ext.rs (hunk 4 above)
Comment (techraed) [reply core-processor/src/ext.rs]: did you mean like this?
https://github.com/gear-tech/gear/pull/4165/commits/d31387c7edd8715c60f2f6f3b8d5032aea76f84d

**Pull request merged**